Floating cursor docs

justinmc opened the pull request:
Explains what a "floating cursor" is in the docs.

Fixes https://github.com/flutter/flutter/issues/132411

justinmc committed:
Explain what floating cursor is where mentioned in the docs

Hixie commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -881,6 +881,8 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
   ///
   /// If this is null, the background cursor is not painted.
+  ///
+  /// @{macro flutter.services.TextInput.floatingCursor}
```
i would just inline the see also in all these places. using a macro means people browsing the source can't easily see what you're trying to inline, and if you ever want to use the macros somewhere that already has a "see also" you'll need to find another solution.

Hixie commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -723,6 +723,13 @@ TextAffinity? _toTextAffinity(String? affinity) {
+/// The floating cursor is an iOS feature in which long pressing and dragging on
+/// the spacebar of the soft keyboard causes the cursor in the text field to
+/// move along with the drag gesture. It's used to precisely position the cursor
+/// in some editable text.
```
that doesn't really help me understand what it is, or why it has multiple colors, etc. would it be possible to describe it in more detail, or maybe generate screenshots showing it (in assets-for-api-docs) and pointing to those?

The audience here is people who have never used iOS. They probably need to know all the various modes it can be in, the various properties that can be used to configure it, and so on.

Hixie replied:
(the writing prompt i usually give myself is, "if someone is looking at this, that means they have a question that they're trying to answer. what is the question? what will follow-up questions be?")

justinmc replied:
Updated, let me know what you think.

justinmc committed:
Inline the macro

justinmc committed:
Include a gif, and reword the docs

justinmc committed:
Include device where screenshot was captured

justinmc commented:
The referenced gif will be added in https://github.com/flutter/assets-for-api-docs/pull/226.

Hixie commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -721,10 +721,27 @@ TextAffinity? _toTextAffinity(String? affinity) {
+/// This feature works out-of-the-box with Flutter. Support is built into
+/// [EditableText].
+///
+/// ![Animation of the native iOS floating cursor feature on an iPhone 14 simulator running iOS 17.0](
```
This is the title of the image, but what we're supposed to provide in the alternative text is a replacement for the image, for someone who can't see it.

I think in this case what I would recommend is to not have any alternative text and just describe the image so that everyone can see the description. Something like, "when the user long-presses the space bar, they are able to drag the cursor to an arbitrary position. The floating cursor snaps to lines as the user moves it vertically, but does not snap to positions between characters horizontally. The floating cursor renders over the text field, while a faint shadow of the cursor is rendered in the text field in the location between characters where the cursor will drop into when released. The shadow cursor is a faint vertical bar with a blob at its bottom, while the floating cursor has the same appearance as a normal cursor (a blue vertical bar)."

Hixie commented:
Yeah, having a screenshot is a huge improvement!

justinmc committed:
Remove the gif and link to the relevant Apple docs

justinmc committed:
Merge branch 'master' into floating-cursor-docs

justinmc commented:
I've removed the screenshot as it's not possible to generate in a script, and I've linked to the relevant Apple docs.

justinmc committed:
Incorporating alt-text-style feedback

justinmc commented:
@Hixie Thanks for the guidance, I've updated it to integrate something close to what you've written. One thing I changed was not mentioning the "blob" at the bottom of the placeholder cursor, which I didn't see when I tried this on an iOS 17 simulator.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -721,10 +721,34 @@ TextAffinity? _toTextAffinity(String? affinity) {
+/// long-presses the spacebar on the iOS virtual keyboard, iOS enters floating
+/// cursor mode where the whole keyboard becomes a trackpad. In this mode,
+/// there are two visible cursors. One, the floating cursor, hovers over the
+/// text, following the user's horizontal movements exactly and snapping to
```
The "floating" cursor follows the user's gesture movements exactly I believe (there's no line-snapping).

justinmc replied:
I tried it again and it does sort of snap. Maybe that's not the best way to call it?

![output](https://github.com/flutter/flutter/assets/389558/9b187a3e-be76-41c1-b176-43563311d3b5)

LongCatIsLooong replied:
Is this in a UIKit app? I don't think Flutter does this and I don't remember UIKit doing this either. It could be new.

LongCatIsLooong replied:
Also both cursors being visible when they are so close to each other is different from the behavior I remembered. Pretty sure it used to hide the gray cursor when it's too close to the blue cursor.

justinmc replied:
This is the Reminders app on an iOS 17 simulator. I see the same behavior on Messages so maybe it's a new iOS 17 thing.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -721,10 +721,34 @@ TextAffinity? _toTextAffinity(String? affinity) {
+/// The state of a "floating cursor" drag on an iOS soft keyboard.
+///
+/// The "floating cursor" cursor-positioning mode is an iOS feature used to
+/// precisely position the caret in some editable text. When the user
```
nit: force touch/long press/two-finger pan also picks up the cursor so maybe:
some editable text using certain touch gestures.  As an example: when the user

justinmc replied:
Good call!

LongCatIsLooong reviewed (approved):
LGTM

justinmc committed:
Mention other ways of moving the floating cursor

justinmc committed:
Merge branch 'master' into floating-cursor-docs
